// vant/lib/utils/event.js
"use strict";

exports.__esModule = true;
exports.on = on;
exports.off = off;
exports.stop = stop;
exports.prevent = prevent;
exports.supportsPassive = void 0;

var _ = require(".");

/* eslint-disable no-empty */

/* eslint-disable getter-return */

/* eslint-disable import/no-mutable-exports */
var supportsPassive = false;
exports.supportsPassive = supportsPassive;

if (!_.isServer) {
  try {
    var opts = {};
    Object.defineProperty(opts, 'passive', {
      get: function get() {
        /* istanbul ignore next */
        exports.supportsPassive = supportsPassive = true;
      }
    });
    window.addEventListener('test-passive', _.noop, opts);
  } catch (e) {}
}

function on(target, event, handler, passive) {
  if (passive === void 0) {
    passive = false;
  }

  if (!_.isServer) {
    target.addEventListener(event, handler, supportsPassive ? {
      capture: false,
      passive: passive
    } : false);
  }
}

function off(target, event, handler) {
  !_.isServer && target.removeEventListener(event, handler);
}

function stop(event) {
  event.stopPropagation();
}

function prevent(event) {
  event.preventDefault();
}
